### DISCUSSION TITLE
Astro compatibility

### DISCUSSION THREAD

**Discussion by jonathanblade** [tauri-apps/tauri#4349] [Q&A]
Is it possible to use `tauri` with [astro](https://astro.build)?

**Comment by senden9** [comment]
It is surely possible to combine them. You _should just_ need to adapt the following values in your `tauri.conf.json`.
```
{
  "build": {
    // The Tauri CLI will execute this command when you run `tauri build`.
    // This is the place to bundle your frontend and place html,css, and js files into the `distDir`.
    "beforeBuildCommand": "npm run build",
    // The Tauri CLI will execute this command when you run `tauri dev`.
    // This should start your development server.
    "beforeDevCommand": "npm run dev",
    "devPath": "http://localhost:8000",
    "distDir": "../dist"
  },
```

**Comment by pricesmith** [comment]
Is there any template or example project? I'd like to use cargo as a package manager but there's ofc no option for initializing with astro.

**Reply by JonasKruckenberg** [reply]
You can't use You can't use cargo to manage astro and it's dependencies as Cargo is strictly Rust focused.

Also, see this template herr if you haven't found it already :)

https://github.com/JonasKruckenberg/tauri-astro-template

**Reply by angel-luis** [reply]
nice thank you, is planned an updated version for the v2 Beta? I think Astro is more suitable than Next.js because is SSG first, while currently Next.js with the App router is server side first. Also Astro is compatible with many libraries/frameworks, while Next.js is closed to React.

**Comment by rodydavis** [comment]
I have it working with tarui and astro

<img width="912" alt="Screenshot 2023-05-25 at 12 34 38 PM" src="https://github.com/tauri-apps/tauri/assets/31253215/29710fae-9441-428d-903a-7cf955e7eb50">

**Comment by CM-IV** [comment]
Yes, Tauri works with Astro - my project uses the SolidJS integration.  After running my final executable from the terminal however, there is a warning/message `Asset 'x' not found; fallback to 'x'.html` which I'm guessing has to do with Astro being for MPAs with multiple html files generated.

**Reply by samrockcash** [reply]
I had the same "Asset 'x' not found; fallback to 'x'.html" issue with a qwik tauri  build. I'm Building with solid for now.

**Comment by lucasdaub** [comment]
The thing to understand is that Tauri will wrap around the pre-existing Astro project. Pay attention to the order :
1. Install NPM
2. Install Rust
3. Install Tauri : ```cargo install tauri --locked```
4. Install Tauri CLI : ```cargo install tauri-cli --locked```
5. Create Astro project : ```npm create astro@latest```

6. Once the Astro project is created, we will use Tauri to integrate into the existing Astro project. To do so, command : ```cargo tauri init```, which creates a minimal Tauri configuration. [Check Tauri doc](https://tauri.app/fr/v1/guides/getting-started/setup/integrate). Tauri will need you to procide a few informations. I give you the base config, please adapt it if you changed the parameters (such as the port or the production files folder name or location) :
  - Name of your app
  - Name of the window
  - Relative path to the production (not dev) files. Per default in Astro the production files are located in the (maybe not currently existing) dist folder. As the path has to be given from the tauri.conf.json file, we will give (per default) ```../dist``` in the prompt
  - The dev server url is per default ```http://localhost:4321/```
  - The frontend dev command in Astro is ```npm run dev```
  - The frontend dev command in Astro is ```npm run build```

7. Once step 6 is done, let's verify that it is working as expected. Let's run the command ```cargo tauri dev```. The dev server will be launched on the url specified (here http://localhost:4321) as a typical Astro project. If you type the url in your browser, you will access it, but it is Astro not Tauri. At the same time, Rust will compile the files (which might take some time on first compilation) and generate the application window. If you have the same window as your web browser, then Tauri is working with Astro ! Here is an example of my Windows :

![image](https://github.com/tauri-apps/tauri/assets/62343580/f8af5978-34b7-49ba-a7b3-4fd8433f179a)

**Reply by evolfo** [reply]
> The thing to understand is that Tauri will wrap around the pre-existing Astro project. Pay attention to the order :
> 
>     1. Install NPM
> 
>     2. Install Rust
> 
>     3. Install Tauri : `cargo install tauri --locked`
> 
>     4. Install Tauri CLI : `cargo install tauri-cli --locked`
> 
>     5. Create Astro project : `npm create astro@latest`
> 
>     6. Once the Astro project is created, we will use Tauri to integrate into the existing Astro project. To do so, command : `cargo tauri init`, which creates a minimal Tauri configuration. [Check Tauri doc](https://tauri.app/fr/v1/guides/getting-started/setup/integrate). Tauri will need you to procide a few informations. I give you the base config, please adapt it if you changed the parameters (such as the port or the production files folder name or location) :
> 
> 
>     * Name of your app
> 
>     * Name of the window
> 
>     * Relative path to the production (not dev) files. Per default in Astro the production files are located in the (maybe not currently existing) dist folder. As the path has to be given from the tauri.conf.json file, we will give (per default) `../dist` in the prompt
> 
>     * The dev server url is per default `http://localhost:4321/`
> 
>     * The frontend dev command in Astro is `npm run dev`
> 
>     * The frontend dev command in Astro is `npm run build`
> 
> 
>     7. Once step 6 is done, let's verify that it is working as expected. Let's run the command `cargo tauri dev`. The dev server will be launched on the url specified (here http://localhost:4321) as a typical Astro project. If you type the url in your browser, you will access it, but it is Astro not Tauri. At the same time, Rust will compile the files (which might take some time on first compilation) and generate the application window. If you have the same window as your web browser, then Tauri is working with Astro ! Here is an example of my Windows :
> 
> 
> ![image](https://private-user-images.githubusercontent.com/62343580/310123871-f8af5978-34b7-49ba-a7b3-4fd8433f179a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDI4NDE2ODksIm5iZiI6MTc0Mjg0MTM4OSwicGF0aCI6Ii82MjM0MzU4MC8zMTAxMjM4NzEtZjhhZjU5NzgtMzRiNy00OWJhLWE3YjMtNGZkODQzM2YxNzlhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMjQlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzI0VDE4MzYyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTc5MTM2YjBmMmYxMzI4MWJlM2RlODk1OTU4ZWI1MTJmYTAxYzE4ZDRlZjFiODQ0M2IzOTcxZTg4YWQ3ZDdjNjMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.0SWkXlAT0AVi_eltKZdTYqE1tNcmevjvh5I27NloBPk)

When following this or the linked guide I keep getting an error when running `cargo tauri dev`, has anyone else experienced this? I just am running the recommended Astro default started template.

```error[E0283]: type annotations needed
   --> /Users/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/plist-1.7.0/src/stream/binary_reader.rs:252:58
    |
252 |                 if value < 0 || value > u64::max_value().into() {
    |                                       -                  ^^^^
    |                                       |
    |                                       type must be known at this point
    |
    = note: multiple `impl`s satisfying `i128: PartialOrd<_>` found in the following crates: `core`, `deranged`:
            - impl PartialOrd for i128;
            - impl<MIN, MAX> PartialOrd<deranged::RangedI128<MIN, MAX>> for i128
              where the constant `MIN` has type `i128`, the constant `MAX` has type `i128`;
help: try using a fully qualified path to specify the expected types
    |
252 |                 if value < 0 || value > <u64 as Into<T>>::into(u64::max_value()) {
    |                                         +++++++++++++++++++++++                ~

error[E0283]: type annotations needed
   --> /Users/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/plist-1.7.0/src/stream/binary_reader.rs:252:58
    |
252 |                 if value < 0 || value > u64::max_value().into() {
    |                                                          ^^^^
    |
note: multiple `impl`s satisfying `_: From<u64>` found
   --> /Users/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/plist-1.7.0/src/integer.rs:91:1
    |
91  | impl From<u64> for Integer {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
   ::: /Users/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/plist-1.7.0/src/value.rs:552:1
    |
552 | impl From<u64> for Value {
    | ^^^^^^^^^^^^^^^^^^^^^^^^
    = note: and more `impl`s found in the following crates: `core`:
            - impl From<u64> for AtomicU64;
            - impl From<u64> for i128;
            - impl From<u64> for u128;
    = note: required for `u64` to implement `Into<_>`
help: try using a fully qualified path to specify the expected types
    |
252 |                 if value < 0 || value > <u64 as Into<T>>::into(u64::max_value()) {
    |                                         +++++++++++++++++++++++                ~

   Compiling core-graphics-types v0.2.0
   Compiling embed-resource v3.0.2
   Compiling objc2-app-kit v0.3.0
   Compiling cargo_toml v0.22.1
For more information about this error, try `rustc --explain E0283`.
error: could not compile `plist` (lib) due to 2 previous errors```

**Comment by angel-luis** [comment]
I just needed to follow this guide into an existing Astro project, works fine: https://tauri.app/v1/guides/getting-started/setup/integrate

**Comment by Toshinaki** [comment]
If anyone is still looking for the answer, here's what I did:
1. follow astro's doc to create a new astro project: https://docs.astro.build/en/install-and-setup/
2. follow tauri's doc to manually add tauri to the project: https://v2.tauri.app/start/create-project/#manual-setup-tauri-cli
    start from step 2 since we already have the project created.

and that's all.

**Comment by Keshav-writes-code** [comment]
Or use my step by step guide

https://keshav.is-a.dev/FreqKnow/programming_stuff/webdev/astro_on_android_with_tauri/
